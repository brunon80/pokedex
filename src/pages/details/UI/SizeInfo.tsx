import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

export type SizeInfoProps = {
  weight: number
  height: number
  baseExpericence: number
}

const SizeInfo: React.FC<SizeInfoProps> = ({ weight, height, baseExpericence }) => {
  return (
    <View style={styles.wrapper} testID="size-info">
      <View style={styles.textWrapper} >
        <Text style={styles.atrubites} testID="height-info">{height ? `${height / 10}m`.replace('.', ',') : '?'}</Text>
        <Text style={styles.label}>Height</Text>
      </View>
      <View style={styles.textWrapper} >
        <Text style={styles.atrubites} testID="weight-info">{weight ? `${weight / 10}kg` : '?'}</Text>
        <Text style={styles.label}>Weight</Text>
      </View>
      <View style={styles.textWrapper} >
        <Text style={styles.atrubites} testID="base-expericence-info">{baseExpericence ? baseExpericence : '?'}</Text>
        <Text style={styles.label}>Base Exp</Text>
      </View>
    </View>
  )
}

export default SizeInfo

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between', 
    paddingHorizontal: 40,
    paddingVertical: 20,
  },
  textWrapper: {
    justifyContent: 'center', 
    alignItems: 'center',
  },
  atrubites: {
    fontWeight: 'bold',
    fontSize: 25
  },
  label: {
    fontSize: 15,
    marginTop: 5
  }
})