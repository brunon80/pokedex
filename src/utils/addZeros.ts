function addZeros(number: string | number): string {
  const convertedNumber = typeof number === 'number' ? number.toString() : number
  const amountOfZeros = {
      [3]: `${number}`,
      [2]: `0${number}`,
      [1]: `00${number}`
  }
  
  return amountOfZeros[convertedNumber?.length] || `${0}`
}

export default addZeros