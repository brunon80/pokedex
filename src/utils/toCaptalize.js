function toCaptalize(string) {
  return `${string?.charAt(0).toUpperCase()}${string?.slice(1)}`.replace('-', ' ')
}

export { toCaptalize }