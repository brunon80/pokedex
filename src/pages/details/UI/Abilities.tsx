import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { toCaptalize } from '../../../utils/toCaptalize';

export type Ability = {
  ability: {
    name: string
  }
}

export type AbilitiesProps = {
  abilities: Ability[]
}

const Abilities: React.FC<AbilitiesProps> = ({ abilities }) => {
  return (
    <View style={styles.container} testID="abilities" >
      <Text style={styles.title}>Abilities</Text>
      <View style={styles.wrapper}>
        {
          abilities?.map(({ ability }) => <Text style={styles.ability} key={ability?.name} testID="ability" >{toCaptalize(ability?.name)}</Text>)
        }
      </View>
    </View>
  )
}

export default Abilities

const styles = StyleSheet.create({
  container: {
    padding: 20,
  },
  wrapper: {
    marginTop: 10,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  title: {
    fontWeight: 'bold',
    textAlign: 'center',
    fontSize: 16
  },
  ability: {
    paddingHorizontal: 10,
    paddingVertical: 5,
    backgroundColor: '#F7F7F7',
    borderRadius: 100,
    overflow: 'hidden',
    alignItems: 'center',
    marginHorizontal: 5
  }
})