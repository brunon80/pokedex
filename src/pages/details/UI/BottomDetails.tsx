import React from 'react'
import { View, StyleSheet } from 'react-native'
import SizeInfo, { SizeInfoProps } from './SizeInfo';
import Abilities, { Ability } from './Abilities';
import Stats from './Stats';
 
type BottomDetailsProps = {
  sizeInfo: SizeInfoProps
  abilites: Ability[]
  pokemonTypeColor: {
    main: string
  }
  stats: Array<{
    base_stat: number,
    stat: {
      name: string
    }
  }>
}


const BottomDetails: React.FC<BottomDetailsProps> = ({ sizeInfo, abilites, pokemonTypeColor, stats }) => {
  return (
    <View  testID="bottom-details" style={styles.wrapper}>
      <SizeInfo height={sizeInfo?.height} weight={sizeInfo?.weight} baseExpericence={sizeInfo?.baseExpericence} />
      <Abilities abilities={abilites} />
      <Stats pokemonTypeColor={pokemonTypeColor} stats={stats} />
    </View>
  )
}

export default BottomDetails;

const styles = StyleSheet.create({
  wrapper: {
    backgroundColor: 'white',
    flexGrow: 1,
    borderTopLeftRadius: 30,
    borderTopRightRadius: 30,
    marginTop: -20
  }
})