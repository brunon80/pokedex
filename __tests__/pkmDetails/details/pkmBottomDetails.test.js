import React from 'react'
import BottomDetails from '../../../src/pages/details/UI/BottomDetails'
import SizeInfo from '../../../src/pages/details/UI/SizeInfo'
import Abilities from '../../../src/pages/details/UI/Abilities'
import { toCaptalize } from '../../../src/utils/toCaptalize';
import { render } from '@testing-library/react-native'

describe('Pkm Details Bottom Tests', () => {

  it('should render BottomDetails correctly', () => {
    const { queryByTestId } = render(<BottomDetails />)
    const component = queryByTestId('bottom-details')
    expect(component).toBeTruthy()
  });

  it('should render SizeInfo correctly', () => {
    const { queryByTestId } = render(<SizeInfo />)
    const component = queryByTestId('size-info')
    expect(component).toBeTruthy()
  });

  it('should render SizeInfo weight correctly', () => {
    const weight = 60

    const { queryByTestId } = render(<SizeInfo weight={weight} />)
    const component = queryByTestId('weight-info')
    expect(component.props.children).toBe('6kg')
  });

  it('should render SizeInfo weight correctly', () => {
    const height = 40

    const { queryByTestId } = render(<SizeInfo height={height} />)
    const component = queryByTestId('height-info')
    expect(component.props.children).toBe('4m')
  });

  it('should render SizeInfo base experience correctly', () => {
    const baseExpericence = 112

    const { queryByTestId } = render(<SizeInfo baseExpericence={baseExpericence} />)
    const component = queryByTestId('base-expericence-info')
    expect(component.props.children).toBe(baseExpericence)
  });

  it('should match snapshot <SizeInfo />', () => {
    const weight = 60
    const height = 40
    const baseExpericence = 112
    const component = render(
      <SizeInfo
        baseExpericence={baseExpericence}
        height={height}
        weight={weight} />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  // Habilidades

  it('should render Abilities correctly', () => {
    const { queryByTestId } = render(<Abilities />)
    const component = queryByTestId('abilities')
    expect(component).toBeTruthy()
  });

  it('should render an array of abilities', () => {

    const abilities = [
      {
        ability: {
          name: 'static'
        }
      },
      {
        ability: {
          name: 'poison pin'
        }
      },
    ]
    
    const { queryAllByTestId } = render(<Abilities abilities={abilities} />)
    const components = queryAllByTestId('ability')
    expect(components.length).toBeGreaterThan(0)

    components.forEach((comp, index) => {
      expect(comp.props.children).toBe(toCaptalize(abilities[index].ability.name))
    })

  })

})