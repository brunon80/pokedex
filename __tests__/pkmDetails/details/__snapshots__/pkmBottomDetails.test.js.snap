// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pkm Details Bottom Tests should match snapshot <SizeInfo /> 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 40,
      "paddingVertical": 20,
    }
  }
  testID="size-info"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 25,
          "fontWeight": "bold",
        }
      }
      testID="height-info"
    >
      4m
    </Text>
    <Text
      style={
        Object {
          "fontSize": 15,
          "marginTop": 5,
        }
      }
    >
      Height
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 25,
          "fontWeight": "bold",
        }
      }
      testID="weight-info"
    >
      6kg
    </Text>
    <Text
      style={
        Object {
          "fontSize": 15,
          "marginTop": 5,
        }
      }
    >
      Weight
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 25,
          "fontWeight": "bold",
        }
      }
      testID="base-expericence-info"
    >
      112
    </Text>
    <Text
      style={
        Object {
          "fontSize": 15,
          "marginTop": 5,
        }
      }
    >
      Base Exp
    </Text>
  </View>
</View>
`;
